File: R/asNettoie.R
asNettoie <- function(Gr) {
  # Le dernier regroupement de Gr est considéré gardé
  # Tous les sous-ensembles qui précèdent sont enlevés
  m <- length(Gr)
  for (k in seq(m, 1)) {
    for (j in Gr[[k]]) {
      for (i in seq_len(k - 1)) {
        if (j %in% Gr[[i]]) {
          Gr[[i]] <- numeric()
        }
      }
    }
  }
  Gr[sapply(Gr,length)==0] <- NULL
  reste <- unlist(Gr[sapply(Gr,length)==1] )
  Gr[sapply(Gr,length)==1] <- NULL
  # Réordonner les groupes selon le rang de leurs variables
  ng <- length(Gr)
  prem <- numeric(ng)
  for (k in seq_len(ng)) {
    Gr[[k]] <- sort(Gr[[k]])
    prem[k] <- Gr[[k]][1]
  }
  
  Gr <- Gr[order(prem)]
  return(list(Gr = Gr, reste = reste))
}

File: Matlab/SCRoF_mars25/asNettoie.m
function [Gr,reste]=asNettoie(Gr)
% Gr=asNnettoie(Gr);
% le dernier regroupement de Gr est considéré gardé
% Tous les sous-ensembles qui précèdent sont enlevés
m=numel(Gr);
for k=m:-1:1
    for j=Gr{k}
        for i=1:k-1
            if any(Gr{i}==j)
                Gr{i}=[];
            end
        end
    end
end
for k=m-1:-1:1
    if isempty(Gr{k})
        Gr(k)=[];
    end
end
reste=[];
for k=numel(Gr):-1:1
    if numel(Gr{k})==1
        reste=[Gr{k} reste];
        Gr(k)=[];
    end
end
% réordonner les groupes selon le rang de leurs variables
ng=numel(Gr);
prem=zeros(ng,1);
for k=1:ng
    Gr{k}=sort(Gr{k});
    prem(k)=Gr{k}(1);
end
[~,oo]=sort(prem);
Gr(:)=Gr(oo);
